### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by daedalus28**
Refactoring, originally prepping for subquery but merging as is because server-side subqueries are being placed on hold (since interactive searches without caching would actually be _less_ efficient than orchestrating the subquery on the client because it would trigger the subquery search to run every time to get the filter results, even if the subquery criteria didn't change).

### CODE REVIEW

**Comment by decrapifier** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  The README has not been updated. Please update the README.
  </td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  You reduced the total lines of code! Awesome! :+1:
  </td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>

  Awesome! Thanks for the well-formatted PR!
  </td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Comment by daedalus28** [comment]
smartprocure/contexture#20 will help us merge this with more confidence since we'll have a more complete test case and can prove this doesn't break anything.

File: src/utils.js [hunk 1]
```diff
@@ -68,9 +61,28 @@ let getRelevantFilters = _.curry((groupCombinator, Path, group) => {
...
+let runTypeFunction = config => async (name, item, search) => {
+  let schema = config.getSchema(item.schema)
+  let fn = F.cascade(
+    [`${item.type}.${name}`, `default.${name}`],
+    config.getProvider(item).types,
+    _.noop
+  )
+  try {
+    return await (search
+      ? fn(item, search, schema, config)
```
Comment (daedalus28) [inline src/utils.js:73]: Let's simplify this by just always passing in all 4 things. This is a breaking change and affects contexture-elasticsearch's `facet` filter function. The issue is that prior to `filter` running, relevantFilters won't be defined which is used to curry the search. We can make the curried search be defined in the run type function thing which will just inspect item on demand, so there won't be any relevant filters if you try to call search from within a filter function. This _might_ even dovetail with upcoming subquery/savedsearch filter work

File: src/utils.js [hunk 2]
```diff
@@ -68,9 +61,28 @@ let getRelevantFilters = _.curry((groupCombinator, Path, group) => {
   return groupCombinator(group, _.compact(relevantFilters))
 })
 
+let runTypeFunction = config => async (name, item, search) => {
+  let schema = config.getSchema(item.schema)
+  let fn = F.cascade(
+    [`${item.type}.${name}`, `default.${name}`],
```
Comment (daedalus28) [inline src/utils.js:67]: Let's simplify this by just killing core level support for "defaults" like this. Individual providers can easily just wrap their type definitions in a `_.map(_.defaults({}))` kind of thing if it's really needed

File: src/utils.js [hunk 3]
```diff
@@ -68,9 +61,28 @@ let getRelevantFilters = _.curry((groupCombinator, Path, group) => {
...
+    config.getProvider(item).types,
+    _.noop
```
Comment (daedalus28) [inline src/utils.js:69]: with the cascade gone, instead of worrying about _.noop, I think we can just use `F.maybeCall`

File: test/index.js [hunk 4]
```diff
@@ -25,15 +25,23 @@ describe('Contexture Core', () => {
         data: {
           value: 1,
         },
+        config: {
+          c: 1,
```
Comment (daedalus28) [inline test/index.js:29]: let's not change the test cases for something this big

File: src/index.js [hunk 5]
```diff
@@ -1,81 +1,54 @@
...
-      await processStep(item => {
+      await walk(
+        // Do all of these in the same traversal
+        overAsync([
+          initNode,
+          flattenLegacyFields,
+          materializePaths,
+          async item => {
+            item._meta.hasValue = await runTypeFunction('hasValue', item)
+            if (item._meta.hasValue && !item.contextOnly) {
```
Comment (daedalus28) [inline src/index.js:45]: Can't find any real use or even any reference to `contextOnly`. Let's just kill it.

File: src/utils.js (hunk 1 above)
Comment (daedalus28) [reply src/utils.js:73]: It is confirmed that nothing is currently using any of the props after schema.

Comment (daedalus28) [reply src/utils.js:73]: also moving search into here will reduce the duplicate get schema call that happens before curry search and in here

**Comment by daedalus28** [comment]
Created smartprocure/monorepo-test#4 as a follow up for the feedback as those are breaking changes. Let's proceed without breaking anything to get subquery/savedsearch support in

**Pull request merged**